### ISSUE TITLE
About

### ISSUE THREAD

**Issue by Cc-Hy** [Xianpeng919/MonoCon#3]
Hello, I tried to train the model, but after 120 epochs, the performance is a lot worse than yours.
The modification is that I used a larger learning rate 0.001 compare to your original 0.000225.
So first I want to ask why the learning rate you choose is so small ( generally contact with the network learning rate around 0.003 to 0.001), do you use pre-training and this is a fine-tuning?
And I want to ask for some idea about the results I got, I think the learning rate would not result in such a large gap.
And I will use your original learning rate to retrain later.
Thanks a lot.

**Comment by Cc-Hy** [comment]
![image](https://user-images.githubusercontent.com/82150240/156491635-f54a3f0b-b226-46f7-9ba9-fd75f5820281.png)
120 epoch results.

**Comment by Cc-Hy** [comment]
Oh, I noticed the lr is the starting lr, not the peak lr.
I will change the lr and try again.

**Comment by rockywind** [comment]
@Cc-Hy @Xianpeng919 
Hi, I trained the model on training set and tested on the validation. The moderate class 3D AP is 17.57, but the paper say it was 19.03.
![4b5afd05bed72801c3c19e8bf00661a](https://user-images.githubusercontent.com/14838781/156564276-09a1006a-c65f-4c77-ba83-f98b435a7e07.png)

**Comment by Xianpeng919** [comment]
@rockywind Did you use the provided config to train your model?

**Comment by Cc-Hy** [comment]
@rockywind  @Xianpeng919
Hello, I tried and get this result, which islower than provided. And I find that the training results seem to be unstable.
I will try another time later.
![image](https://user-images.githubusercontent.com/82150240/156695157-5d950b23-cf57-477d-a144-ae9d9db7a5d2.png)

**Comment by Cc-Hy** [comment]
![image](https://user-images.githubusercontent.com/82150240/156695333-1124ed97-36f9-4c69-8ed1-570f51aa1bcd.png)
![image](https://user-images.githubusercontent.com/82150240/156695394-ace69454-cd38-476c-a382-055c8adbe1e5.png)
Another thing, I modified the lr to be 0.0001, and the max lr should be 0.0003, but I still got lr like 0.005.
Did I misunderstand the config settings?
Thanks.

**Comment by rockywind** [comment]
@Xianpeng919 Yes, I use the default config.
[https://github.com/Xianpeng919/MonoCon/blob/main/monocon/configs/monocon/monocon_dla34_200e_kitti.py](url)

**Comment by rockywind** [comment]
@Xianpeng919 
I trained the model the second time. The result is below.
````
3D APR40: 23.7064, 17.7595, 14.9525
````

**Comment by Xianpeng919** [comment]
@rockywind I'll double check and get back to you asap.

**Comment by Cc-Hy** [comment]
@Xianpeng919 
Hello, how to modify the cfg file if I want to train with the trainval set and get test results?

**Comment by rockywind** [comment]
@Xianpeng919 
 I load the pretrained model and train the model. The result is below.
````
3D APR40: 24.2891, 18.0508, 15.2171
````

**Comment by kaixinbear** [comment]
Hello, I train the model with command `CUDA_VISIBLE_DEVICES=0  python ./tools/train.py configs/monocon/monocon_dla34_200e_kitti.py`
 without any modification but the performance is rather lower. The permance get its peak at 120 epoch and get lower and lower util 0.
The result at 120 epoch:
```
Car AP@0.70, 0.70, 0.70:
3d   AP:16.5400, 12.2644, 10.5623
```
The result at 200 epoch:
```
Car AP@0.70, 0.70, 0.70:
3d   AP:0.0000, 0.0000, 0.0000
```
[trainging log could be seen here](https://paste.ubuntu.com/p/4VPsRqzqPh/)
What I should do to get a rather normal result ? @Xianpeng919

**Comment by Xianpeng919** [comment]
@rockywind We have tested our released checkpoints in multiple GPUs. The result is 26.33 | 19.03 | 16.00, same as the result in the readme. Not sure what the problem is here. You might provide me with your log so that I can help you check the details.

**Comment by Xianpeng919** [comment]
@Cc-Hy You may replace the training split with the trainval split in the config

**Comment by Xianpeng919** [comment]
@kaixinbear Your dimension branch exploded during training. We did observe this during our experiments. The dimension-aware loss is a little bit unstable. You can restart your training from the un-exploded ckpts.

**Comment by kaixinbear** [comment]
Thanks for your kindly reply! I will try later

**Comment by ganyz** [comment]
> @kaixinbear Your dimension branch exploded during training. We did observe this during our experiments. The dimension-aware loss is a little bit unstable. You can restart your training from the un-exploded ckpts.

Hello author，I resume my training from the un-exploded ckpts, but it still explodes in the follwing epochs. Have you met this phenomenon？ Should i turn down my lr ? 
Thanks!

**Comment by rockywind** [comment]
@Xianpeng919 
I tested the released checkpoint. The result is the same as the readme. When I retrained the model, the result was lower than the readme.
[20220302_134704.log](https://github.com/Xianpeng919/MonoCon/files/8197749/20220302_134704.log)

**Comment by excitohe** [comment]
> @Xianpeng919 I tested the released checkpoint. The result is the same as the readme. When I retrained the model, the result was lower than the readme. [20220302_134704.log](https://github.com/Xianpeng919/MonoCon/files/8197749/20220302_134704.log)

Hi, have your tried multi-gpu training or are you still use single gpu training? I retrained with 4-gpu and get lower results than the readme.
https://paste.ubuntu.com/p/CtJH9Hk52F/

**Comment by Xianpeng919** [comment]
@ganyz You can restart the training from scratch.

**Comment by Xianpeng919** [comment]
@rockywind I double checked your log, the config looks good to me. I'll double check the code. You can also try another random seed to train again to see the performance.

**Comment by rockywind** [comment]
@Xianpeng919 
OK， thanks a lot!

**Comment by Cc-Hy** [comment]
@rockywind @ganyz  @kaixinbear  @Xianpeng919 
I find that during the training, there will be several epochs whose performance is extremely low(close to 0), and the performance from the last epoch may differ by more than 10 points.
Did you meet this situation?

**Comment by Cc-Hy** [comment]
Epoch 112
![image](https://user-images.githubusercontent.com/82150240/157621972-3885db61-e28b-4e7d-9555-b6874a64d83e.png)
Epoch 115
![image](https://user-images.githubusercontent.com/82150240/157622024-a20c061d-181d-45d2-85e6-6a6c383925d8.png)

**Comment by Cc-Hy** [comment]
Tried another time, and the best performance is as follow:
![image](https://user-images.githubusercontent.com/82150240/157657541-7b3315c3-52c9-4b1b-b24a-7591261dda17.png)

**Comment by djp1235a** [comment]
I conduct 3 experiments with different seeds, and the best performance is 17.80. Besides, results are not reproducible with the same seed and deterministic==True in the codebase.

**Comment by excitohe** [comment]
I retrained twice and got 16.20 on the GTX1080Ti and 16.80 on the Titan V. It seems that no one in the issue can retrain more than 18.00, makes me frustrated....  =_=!

**Comment by excitohe** [comment]
> @rockywind @ganyz @kaixinbear @Xianpeng919 I find that during the training, there will be several epochs whose performance is extremely low(close to 0), and the performance from the last epoch may differ by more than 10 points. Did you meet this situation?

It's normal. Mono3D performance is always unstable. Just pay attention to the last few checkpoint eval results. 0.0

**Comment by djp1235a** [comment]
@excitohe  I konw that the Mono3D performance is always unstable. But results are reproducible with the same seed and deterministic==True in the Monodle  codebase. I'm just wondering why nondeterministic algorithms appear when using mmdet reimplementation.

**Comment by excitohe** [comment]
@djp1235a 
[Unified reply from OpenMMLab](https://www.zhihu.com/question/453511684/answer/2032751332)

**Comment by Xianpeng919** [comment]
@excitohe @djp1235a @Cc-Hy I'm re-training the model based on the released code using different GPUs. I'll share with you  the log in this thread once the result is out.

**Comment by Cc-Hy** [comment]
@Xianpeng919 
Hello, the figures in the front page look good, can you tell me how to draw them?
![image](https://user-images.githubusercontent.com/82150240/157814113-d67e89de-524d-4d08-88bf-261581dfe5cf.png)

**Comment by Xianpeng919** [comment]
@Cc-Hy You can refer to mmdet3d's visualization scripts. Their scripts are very helpful.

**Comment by Cc-Hy** [comment]
@Xianpeng919 
Hello, I tried to add "--show" arg in test.py, and I tried to directly use the mono_det_demo.py.
But both of them can not work properly.
Can you tell me which script do you use? And do I need to do some modifications?

**Comment by Xianpeng919** [comment]
@Cc-Hy You can do inference you model first and revise the show_results function in the mmdet3d.core.visualizer

**Comment by excitohe** [comment]
@Xianpeng919  Have you finished your retraining results yet? Looking forward to your train log file. ^_^

**Comment by Xianpeng919** [comment]
@excitohe Hi, I was travelling last weekend. Please check this [log](https://drive.google.com/file/d/1rhXil4cLa8W8kPIRG2ISX98Rfm8RUk0x/view?usp=sharing) for more details. I also attach the [ckpt ](https://drive.google.com/file/d/1iRkf8lHRlELKJGPQ5ukoxJLhGoSxfCGn/view?usp=sharing) here in case you need it. Please run *_car.py config for inference.

**Comment by Cc-Hy** [comment]
Tried again:
![image](https://user-images.githubusercontent.com/82150240/158717179-07f5b00a-04c5-4829-9c0e-762bcaaa283f.png)

**Comment by excitohe** [comment]
Hi, I migrate monocon into latest mmdet3d in plugin_dir manner, and try again with only_car with your latest updated config in 4GPU.  
```
Car AP40@0.70, 0.70, 0.70:
bbox AP40:96.3800, 90.3432, 80.7128
bev  AP40:29.0449, 22.2251, 19.4256
3d   AP40:21.4625, 16.1725, 14.3990
aos  AP40:95.73, 89.51, 79.49
```
Attach the training log:
https://paste.ubuntu.com/p/HyryFkZspc/

Can you see where is the problem？Thank you so much and keep in touch. ^_^

I will reconfigure your original environment and test again with single GPU...

**Comment by excitohe** [comment]
@Cc-Hy Hi, Is this your recent result in only_car config? It looks like we're about the same...

**Comment by Cc-Hy** [comment]
> @Cc-Hy Hi, Is this your recent result in only_car config? It looks like we're about the same...

No, these are 3 class results.
I'm training with Car only now.

**Comment by Cc-Hy** [comment]
Car only
![image](https://user-images.githubusercontent.com/82150240/158930878-5e48bb59-0a05-4618-b96a-6938c4cf5dd3.png)

**Comment by kaixinbear** [comment]
@Cc-Hy @Xianpeng919 @ganyz 

Could you please tell me how to solve this model collapse problem? By turn down lr or change random seed? 
I have tried many times but the AP drops to 0 at about 120 epoch.

**Comment by Cc-Hy** [comment]
> 

If you always meet this problem, you can modify the dimension loss with L1 loss only, L = |gt - pred|.
And then the dimension loss will never explode.
@kaixinbear

**Comment by gervaisi** [comment]
@Xianpeng919 
I want to use the model with mono_det_demo.py but it asks me an annotation file, where can i find it ? I precise that i've already trained the model

**Comment by ShiAngWang** [comment]
@Xianpeng919 @rockywind @Cc-Hy @excitohe @djp1235a @ganyz @gervaisi  @kaixinbear  @djp1235a  Hello, I trained monocon from scratch for 200 epochs with default setting in configs/monocon/monocon_dla34_200e_kitti.py without any change, here is my result:
![image](https://user-images.githubusercontent.com/26656883/209459530-feb5240a-3e1a-4cf6-be4c-6994e36551e0.png)

First, the result for car does not achieve the result in paper:26.33 19.01 15.98, and what makes me more confused, the result seems to be like the result of class-specific in paper rather than class-agnostic because the Ped result is pretty high while the Cyc result is rather low, here is the result in paper:
![image](https://user-images.githubusercontent.com/26656883/209459588-025f9256-3048-47a6-aa69-b232efcb5a38.png)

So, maybe, I guess the default setting is class-specific? Do you guys know the default setting is class-agnostic or class-specific? And if it is class-specific, how to change the setting to class-agnostic to get a better result?

It will be a great favor if you can give me any reply , thank you very much!!!

**Comment by FlyingAnt2018** [comment]
> @Cc-Hy @Xianpeng919 Hi, I trained the model on training set and tested on the validation. The moderate class 3D AP is 17.57, but the paper say it was 19.03. ![4b5afd05bed72801c3c19e8bf00661a](https://user-images.githubusercontent.com/14838781/156564276-09a1006a-c65f-4c77-ba83-f98b435a7e07.png)

Hi, i got AP 19.0217 by setting "cfg.SEED = 1903919922 "
